### ISSUE TITLE
Quantize_training_test fails with matmul operation on Ubuntu 16.04

### ISSUE THREAD

**Issue by namrata-ibm** [tensorflow/tensorflow#12136]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**:  'v1.2.1-0-gb4957ff', '1.2.1'
- **Python version**:  2.7.12
- **Bazel version (if compiling from source)**: 0.4.5
- **CUDA/cuDNN version**: No GPU
- **GPU model and memory**: No GPU
- **Exact command to reproduce**: bazel test -c opt //tensorflow/python:quantize_training_test

### The problem
The `testQuantizedSaveRestore` from  `//tensorflow/python:quantize_training_test` is failing on s390x while importing graph [here.](https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/python/training/quantize_training_test.py#L72) 

The error message shown is : 
`ValueError: Shapes must be equal rank, but are 0 and 2 for 'a/Min/AssignValue' (op: 'Assign') with input shapes: [], [1,1].
`
The check for this failure is at [Merge function](https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/core/framework/shape_inference.cc#L374) . This Merge is called by `Assign` Op kernel. 

I tried changing the `math_ops.matmul` operation from the above test to `math_ops.subtract/minimum/multiply`. With these operations the test passes after removing asserts for `'a/Min/Variable:0'` or `'b/read/Max/Variable:0'` etc. I suppose other operations do not create these tensors. 

Could anyone please provide some inputs on this failure? I am not aware about the computations that are happening when graph is imported with the `matmul` operation.

### Source code / logs
```
.E.
======================================================================
ERROR: testQuantizedSaveRestore (__main__.PywrapQuantizeTrainingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/test/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/quantize_training_test.runfiles/org_tensorflow/tensorflow/python/training/quantize_training_test.py", line 73, in testQuantizedSaveRestore
    _ = importer.import_graph_def(result, name='')
  File "/home/test/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/quantize_training_test.runfiles/org_tensorflow/tensorflow/python/framework/importer.py", line 409, in import_graph_def
    ops.set_shapes_for_outputs(op)
  File "/home/test/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/quantize_training_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 1873, in set_shapes_for_outputs
    shapes = shape_func(op)
  File "/home/test/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/quantize_training_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 1823, in call_with_requiring
    return call_cpp_shape_fn(op, require_shape_fn=True)
  File "/home/test/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/quantize_training_test.runfiles/org_tensorflow/tensorflow/python/framework/common_shapes.py", line 610, in call_cpp_shape_fn
    debug_python_shape_fn, require_shape_fn)
  File "/home/test/.cache/bazel/_bazel_test/24685d064c07f7346b48c2d13ec3ad69/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/quantize_training_test.runfiles/org_tensorflow/tensorflow/python/framework/common_shapes.py", line 676, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Shapes must be equal rank, but are 0 and 2 for 'a/Min/AssignValue' (op: 'Assign') with input shapes: [], [1,1].
```

**Comment by ali01** [comment]
Could you please provide more information about what exactly you are running? What steps should we follow to reproduce?

**Comment by namrata-ibm** [comment]
@ali01, I m running : `bazel test -c opt //tensorflow/python:quantize_training_test`.
As per above log, ` testQuantizedSaveRestore` fails with error `Shapes must be equal rank, but are 0 and 2 for 'a/Min/AssignValue' (op: 'Assign') with input shapes: [], [1,1].`

I tried printing the graph_def object in `tensorflow/python/framework/importer.py`. It is similar to x86:

```
(Pdb) p graph_def
node {
  name: "save/RestoreV2/shape_and_slices"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
          dim {
            size: 1
          }
        }
        string_val: ""
      }
    }
  }
}
.
.
node {
  name: "save/restore_all"
  op: "NoOp"
  input: "^save/Assign"
  input: "^save/Assign/_1"
  input: "^save/Assign/_3"
  input: "^save/Assign/_5"
  input: "^save/Assign/_7"
}
.
.
library {
}
versions {
  producer: 22
}

```

So I further printed 2 things:
1.  `ops` in [importer.py](https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/python/framework/importer.py#L407)
using:   
```
+        print("op.type:", op.type)
         ops.set_shapes_for_outputs(op)
```

 
 2.  rank in [shape_inference.cc](https://github.com/tensorflow/tensorflow/blob/v1.2.1/tensorflow/core/framework/shape_inference.cc#L371) using
```
+  printf("\nRank(s0):%d\n", Rank(s0));
+  printf("Rank(s1):%d\n", Rank(s1));
```

 which gives:

```
op.type: Const
op.type: Const
op.type: Const
op.type: Const
op.type: Const
op.type: RestoreV2
Rank(s0):1
Rank(s1):1

op.type: VariableV2
op.type: Assign
op.type: NoOp
op.type: Identity
op.type: Identity
op.type: Const
op.type: Assign
Rank(s0):2
Rank(s1):2

op.type: NoOp
op.type: Const
op.type: MatMul
op.type: Const
op.type: Const
op.type: Const
op.type: Rank
op.type: Range
op.type: Min
op.type: Max
op.type: VariableV2
op.type: IsVariableInitialized
op.type: Switch
op.type: Const
op.type: Sub
op.type: Sub
op.type: Mul
op.type: Sub
op.type: Merge
op.type: Assign
Rank(s0):0
Rank(s1):2
```

And fails here while on x86, the ranks are both 0 at this point.
So I am unable to understand the exact root cause of this on s390x(which is big endian). Could you please provide inputs in understanding how the `matmul` operation works while importing a graph.

**Comment by namrata-ibm** [comment]
This was related to the numpy bug and resolved by #12963.
Closing this.
